code/artificial_intelligence/src/ant_colony_optimization/aco_tsp.cpp

```diff
@@ -0,0 +1,170 @@
+
+bool reachStagnation (int N, vector<double>& len, vector<double>& previousLen)
+{   // Return true if we have reached stagnation
+    // We have reached stagnation if "len" matrix is equal to "previousLen"
+    for (int i = 0; i < N; i++)
+        for (int j = 0; j < N; j++)
+            if (len[i] != previousLen[j]) 
+                return false;
+    return true;
+}
+
+int main()
+{
+    int N;    // Number of vertices
+    vector<vector<double>> distance, pheromone;    // Adjacency matrices of weighted edges and pheromone values on them
+    vector<vector<int>> parents;     // Parents of vertices in every tour => stores the path for every ant.
+    vector<double> len;              // Lengths of each ant's tour
+    vector<double> previousLen;      // Stores the length of each ant's tour in a previous iteration
+    vector<double> v0;               // Empty vector of double
+    vector<int> i0;                  // Empty vector of int
+    double lowestLen = double (RAND_MAX); // Stores current lowest cost
+    vector<int> cheapestPath; // Stores the cheapest path
+    double x;
+    cin >> N;
+    int maxIter = 30000;
+    srand(time(NULL));
+    for (int i = 0; i < N; i++)
+    {
+        distance.push_back(v0);
+        pheromone.push_back(v0);
+        parents.push_back(i0);
+        len.push_back(0);
+        previousLen.push_back(0);
+        for (int j = 0; j < N; j++)
+        {
+            cin >> x;
+            distance[i].push_back(x);     // Stores adjacency matrices of weighted edges
+            if (i != j) pheromone[i].push_back(0.1);     // Initialize the level of pheromones
+            else pheromone[i].push_back(0);        // Since there is no loop => there is no pheromone
+            parents[i].push_back(-1);       // Each node has no parents yet => notate the absence of parent as -1
+        }
+    }
+    bool stagnation = false;      // We don't have a stagnation yet
+    while (!stagnation && maxIter--)
+    {
+        for (int k = 0; k < N; k++)     // For each  of N ants we need to find their tours
+            len[k] = lkAfterCompleteTour(k,N,distance,pheromone,parents);     // Build the tour for k-th ant and returns the length of this tour
+        updatePheromone(N, pheromone, parents, len);     // Update pheromone table
+        stagnation = reachStagnation(N, len, previousLen);     // Return true if we have reached stagnation
+        for(int k = 0; k < N;k++)
+            if(len[k] < lowestLen) // Stores the best path seen so far
+            {
+                lowestLen = len[k];
+                cheapestPath = parents[k];
+            }
+        if (!stagnation && maxIter) updateParentsAndLength(N, parents, len, previousLen);     // Empty "parents" and update previousLen
+    }
+    int k = 0;
+    // Prints out the shortest tour
+    cout<< "\nTour is: ";
+    for (int i = 0; i<N; i++)
+    {
+        cout << k << "->";
+        k = cheapestPath[k];
+    }
+    cout << 0 <<"\nCost is: " << lowestLen<<endl;
```

Comment:
```suggestion
    cout << 0 << "\nCost is: " << lowestLen << '\n';
```
Make sure other lines also have spacing around operators.